File: polyglot/java/binary_diff.java
package polyglot.java;

import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.security.MessageDigest;
import java.security.cert.CertificateException;
import javax.crypto.spec.SecretKeySpec;

/**
 * Firmware Binary Diff Tool for fwxray.
 * Compares two firmware images and surfaces: new binaries, flipped config flags,
 * added certs, and shifted entropy regions.
 */
public class binary_diff {

    // --- Configuration Constants ---
    private static final int ELF_MAGIC = 0x464C457F;       // "ELF\0" (little-endian)
    private static final int PE_MAGIC = 0x0100;             // PE32 header signature
    private static final int FAT_MAGIC = 0x46415421;        // FAT32 magic
    private static final int CONFIG_SECTION_SIZE = 8192;    // Typical config section size

    // --- Main Entry Point (Runnable Demo) ---
    public static void main(String[] args) throws Exception {
        if (args.length < 2) {
            System.out.println("Usage: java binary_diff <image1> <image2>");
            System.exit(0);
        }

        Path img1 = Paths.get(args[0]);
        Path img2 = Paths.get(args[1]);

        if (!Files.exists(img1)) {
            throw new FileNotFoundException("Image 1 not found: " + img1);
        }
        if (!Files.exists(img2)) {
            throw new FileNotFoundException("Image 2 not found: " + img2);
        }

        System.out.println("=== fwxray Binary Diff ===");
        System.out.println("Image 1: " + img1.toAbsolutePath());
        System.out.println("Image 2: " + img2.toAbsolutePath());
        System.out.println();

        // Step 1: Load and parse both images
        FirmwareImage imgA = new FirmwareImage(img1);
        FirmwareImage imgB = new FirmwareImage(img2);

        // Step 2: Compute high-level diffs
        DiffReport report = computeDiff(imgA, imgB);

        // Step 3: Print results
        printReport(report);
    }

    // --- Core Diff Logic ---
    private static DiffReport computeDiff(FirmwareImage a, FirmwareImage b) {
        DiffReport r = new DiffReport();

        // 1. New binaries (files present in B but not A)
        Set<String> filesA = new HashSet<>(a.files.keySet());
        Set<String> filesB = new HashSet<>(b.files.keySet());
        
        for (String f : filesB) {
            if (!filesA.contains(f)) {
                r.newBinaries.add(f);
            }
        }

        // 2. Flipped config flags
        r.configFlags = compareConfig(a, b);

        // 3. Added certificates
        r.certsAdded = findCertDifferences(a, b);

        // 4. Shifted entropy regions
        r.entropyShifts = analyzeEntropyShifts(a, b);

        return r;
    }

    // --- FirmwareImage: Container for parsed image data ---
    static class FirmwareImage {
        Path path;
        long size;
        String magic;
        int headerType;
        Map<String, byte[]> sections = new LinkedHashMap<>();
        Set<String> files = new HashSet<>();

        FirmwareImage(Path p) throws IOException {
            this.path = p;
            this.size = Files.readAllBytes(p).length;
            
            // Detect magic and parse header
            byte[] header = Files.readAllBytes(Files.probeDirectory(p));
            if (header.length < 4) return;

            int m = Integer.reverseBytes(header[0] | header[1] << 8 | 
                                          header[2] << 16 | header[3] << 24);
            
            this.magic = new String(header, 0, 4);
            this.headerType = m;

            // Parse sections based on magic
            parseSections(header);
        }

        private void parseSections(byte[] header) {
            if (header.length < 64) return;

            switch (this.magic) {
                case "ELF":
                    parseElfSections(header);
                    break;
                case "PE32\0":
                    parsePeSections(header);
                    break;
                default:
                    // Generic section parsing for unknown formats
                    extractGenericSections(header);
            }
        }

        private void parseElfSections(byte[] header) {
            if (header.length < 64) return;
            
            int e_shoff = Integer.reverseBytes(header[0x2A]);
            int e_shentsize = Integer.reverseBytes(header[0x38]);
            int e_shnum = Integer.reverseBytes(header[0x39]);

            for (int i = 0; i < e_shnum && i * e_shentsize + 64 < header.length; i++) {
                int sh_offset = Integer.reverseBytes(header[0x2A + i * e_shentsize + 0x18]);
                if (sh_offset > 0) {
                    sections.put("elf_section_" + i, 
                        Arrays.copyOfRange(header, sh_offset, Math.min(sh_offset + 64, header.length)));
                }
            }
        }

        private void parsePeSections(byte[] header) {
            if (header.length < 128) return;
            
            int peOffset = Integer.reverseBytes(header[0x3C]);
            if (peOffset > 0 && peOffset + 64 < header.length) {
                sections.put("pe_section_0", 
                    Arrays.copyOfRange(header, peOffset, Math.min(peOffset + 64, header.length)));
            }
        }

        private void extractGenericSections(byte[] header) {
            // Look for common section markers in the binary
            String text = new String(header);
            
            if (text.contains(".text") || text.contains(".data")) {
                sections.put("generic_sections", Arrays.copyOfRange(header, 0, Math.min(128, header.length)));
            }

            // Extract any embedded files or archives
            extractEmbeddedFiles(header);
        }

        private void extractEmbeddedFiles(byte[] header) {
            // Look for ZIP/PEM signatures indicating embedded content
            int zipSig = Integer.reverseBytes(header[0] | header[1] << 8 | 
                                             header[2] << 16 | header[3] << 24);
            
            if (zipSig == 0x504B0304) { // ZIP local file header
                sections.put("embedded_zip", Arrays.copyOfRange(header, 0, Math.min(256, header.length)));
            }
        }
    }

    // --- DiffReport: Container for all differences found ---
    static class DiffReport {
        List<String> newBinaries = new ArrayList<>();
        Map<String, String> configFlags = new LinkedHashMap<>();
        List<CertDiff> certsAdded = new ArrayList<>();
        List<EntropyShift> entropyShifts = new ArrayList<>();

        public void print() {
            System.out.println("--- New Binaries ---");
            if (newBinaries.isEmpty()) {
                System.out.println("  None detected.");
            } else {
                for (String f : newBinaries) {
                    System.out.println("  + " + f);
                }
            }

            System.out.println("\n--- Config Flags ---");
            if (configFlags.isEmpty()) {
                System.out.println("  No flag changes detected.");
            } else {
                for (Map.Entry<String, String> e : configFlags.entrySet()) {
                    System.out.println("  ~ " + e.getKey() + " = " + e.getValue());
                }
            }

            System.out.println("\n--- Certificates ---");
            if (certsAdded.isEmpty()) {
                System.out.println("  No new certificates detected.");
            } else {
                for (CertDiff c : certsAdded) {
                    System.out.println("  + " + c.subject + " (" + c.fingerprint.substring(0,8) + "...") ;
                }
            }

            System.out.println("\n--- Entropy Shifts ---");
            if (entropyShifts.isEmpty()) {
                System.out.println("  No entropy region shifts detected.");
            } else {
                for (EntropyShift s : entropyShifts) {
                    System.out.println("  >> " + s.from + " -> " + s.to);
                }
            }
        }
    }

    // --- Config Flag Comparison ---
    private static Map<String, String> compareConfig(FirmwareImage a, FirmwareImage b) {
        Map<String, String> changes = new LinkedHashMap<>();

        // Look for common config formats in sections
        byte[] dataA = extractConfigData(a);
        byte[] dataB = extractConfigData(b);

        if (dataA != null && dataB != null) {
            try {
                // Try JSON parsing first
                String jsonA = new String(dataA, "UTF-8");
                String jsonB = new String(dataB, "UTF-8");

                if (jsonA.trim().startsWith("{") || jsonB.trim().startsWith("{")) {
                    changes = parseJsonConfigDiffs(jsonA, jsonB);
                } else {
                    // Binary blob comparison for config sections
                    changes = compareBinaryConfigs(dataA, dataB);
                }
            } catch (Exception e) {
                // Fallback: simple XOR diff for binary blobs
                changes = compareBinaryConfigs(dataA, dataB);
            }
        }

        return changes;
    }

    private static byte[] extractConfigData(FirmwareImage img) {
        if (img.sections.isEmpty()) return null;

        // Look for config-like sections (typically contain "config", "cfg", etc.)
        for (byte[] section : img.sections.values()) {
            try {
                String text = new String(section);
                if (text.toLowerCase().contains("config") || 
                    text.toLowerCase().contains("cfg")) {
                    return section;
                }
            } catch (Exception e) {}
        }

        // Fallback: first 4KB of image
        return img.size > 0 ? Arrays.copyOfRange(Files.readAllBytes(img.path), 0, Math.min(4096, (int)Math.min(img.size, 4096))) : null;
    }

    private static Map<String, String> parseJsonConfigDiffs(String jsonA, String jsonB) {
        Map<String, String> changes = new LinkedHashMap<>();

        try {
            // Simple JSON diff (no external library)
            String[] linesA = jsonA.split("\n");
            String[] linesB = jsonB.split("\n");

            for (int i = 0; i < Math.max(linesA.length, linesB.length); i++) {
                String lineA = linesA[i].trim();
                String lineB = linesB[i].trim();

                if (!lineA.equals(lineB)) {
                    // Detect flag changes
                    if (lineA.contains("true") && !lineB.contains("true")) {
                        changes.put("flag_" + i, "false");
                    } else if (!lineA.contains("true") && lineB.contains("true")) {
                        changes.put("flag_" + i, "true");
                    }
                }
            }
        } catch (Exception e) {}

        return changes;
    }

    private static Map<String, String> compareBinaryConfigs(byte[] a, byte[] b) {
        Map<String, String> changes = new LinkedHashMap<>();

        if (a == null || b == null) return changes;

        // Simple XOR-based diff for binary config blobs
        int minLen = Math.min(a.length, b.length);
        
        for (int i = 0; i < minLen; i++) {
            if ((a[i] ^ b[i]) != 0) {
                // Check if this looks like a boolean flag change
                if (i % 8 == 0 && 
                    (a[i] & 0x01) == 0 && (b[i] & 0x01) != 0) {
                    changes.put("bit_" + i, "flipped");
                }
            }
        }

        return changes;
    }

    // --- Certificate Difference Detection ---
    private static List<CertDiff> findCertDifferences(FirmwareImage a, FirmwareImage b) {
        List<CertDiff> added = new ArrayList<>();

        byte[] dataA = extractConfigData(a);
        byte[] dataB = extractConfigData(b);

        if (dataA == null || dataB == null) return added;

        try {
            // Look for X.509 certificate headers
            int pemStart = findPemHeader(dataA, 1024);
            
            if (pemStart >= 0) {
                // Extract and parse certificates from both images
                List<CertInfo> certsA = extractCerts(dataA, pemStart);
                List<CertInfo> certsB = extractCerts(dataB, pemStart);

                for (CertInfo c : certsB) {
                    boolean foundInA = false;
                    for (CertInfo ca : certsA) {
                        if (ca.fingerprint.equals(c.fingerprint)) {
                            foundInA = true;
                            break;
                        }
                    }
                    
                    if (!foundInA) {
                        added.add(new CertDiff(
                            c.subject, 
                            c.issuer, 
                            c.fingerprint, 
                            "new"
                        ));
                    }
                }
            }
        } catch (Exception e) {}

        return added;
    }

    private static int findPemHeader(byte[] data, int maxLen) {
        // Look for PEM header markers
        String pemStart = new String(data).indexOf("-----BEGIN CERTIFICATE-----");
        
        if (pemStart >= 0 && pemStart < maxLen) {
            return pemStart;
        }

        // Also check for DER-encoded certificates
        int derSig = Integer.reverseBytes(data[0] | data[1] << 8 | 
                                         data[2] << 16 | data[3] << 24);
        
        if (derSig == 0x3082 || derSig == 0x3081) { // ASN.1 SEQUENCE length prefix
            return 0;
        }

        return -1;
    }

    private static List<CertInfo> extractCerts(byte[] data, int offset) throws Exception {
        List<CertInfo> certs = new ArrayList<>();

        String pemStartMarker = "-----BEGIN CERTIFICATE-----";
        int pos = 0;

        while (pos < data.length - pemStartMarker.length()) {
            int idx = new String(data).indexOf(pemStartMarker, pos);
            
            if (idx == -1) break;

            // Extract certificate content between markers
            int endIdx = new String(data).indexOf("-----END CERTIFICATE-----", idx);
            
            if (endIdx > 0) {
                byte[] certData = Arrays.copyOfRange(data, idx + pemStartMarker.length(), 
                                                      endIdx - idx - pemStartMarker.length());

                // Compute fingerprint for quick comparison
                MessageDigest md = MessageDigest.getInstance("SHA-256");
                byte[] hash = md.digest(certData);
                
                certs.add(new CertInfo(
                    new String(hash).substring(0, 8),
                    "Unknown",
                    new String(hash)
                ));

                pos = endIdx + pemStartMarker.length();
            } else {
                break;
            }
        }

        return certs;
    }

    // --- Entropy Region Shift Analysis ---
    private static List<EntropyShift> analyzeEntropyShifts(FirmwareImage a, FirmwareImage b) {
        List<EntropyShift> shifts = new ArrayList<>();

        byte[] dataA = extractConfigData(a);
        byte[] dataB = extractConfigData(b);

        if (dataA == null || dataB == null) return shifts;

        // Calculate entropy for both images
        double entA = calculateEntropy(dataA);
        double entB = calculateEntropy(dataB);

        // Look for regions with significant entropy changes
        int windowSize = 256;
        
        if (dataA.length >= windowSize * 4 && dataB.length >= windowSize * 4) {
            for (int i = 0; i < Math.min(dataA.length, dataB.length) - windowSize * 3; i += windowSize) {
                byte[] chunkA = Arrays.copyOfRange(dataA, i, i + windowSize);
                byte[] chunkB = Arrays.copyOfRange(dataB, i, i + windowSize);

                double entChunkA = calculateEntropy(chunkA);
                double entChunkB = calculateEntropy(chunkB);

                // Detect significant entropy shift (> 2.0 bits difference)

File: polyglot/rust/binary_diff.rs
use std::collections::{HashMap, HashSet};
use std::fs;
use std::path::Path;
use std::io::{self, Read};
use sha2::{Sha256, Digest};

/// Represents a detected binary file type
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum BinaryType {
    Unknown,
    Elf32,
    Elf64,
    Pe32,
    Pe64,
    Fat,
}

impl Default for BinaryType {
    fn default() -> Self {
        BinaryType::Unknown
    }
}

/// Metadata extracted from a firmware image
#[derive(Debug)]
pub struct FirmwareImage {
    pub path: Option<PathBuf>,
    pub data: Vec<u8>,
    pub binary_type: BinaryType,
    pub checksum: String,
    pub entropy: f64,
    pub config_regions: Vec<ConfigRegion>,
}

/// A region in the firmware that may contain configuration flags
#[derive(Debug)]
pub struct ConfigRegion {
    pub offset: u64,
    pub size: usize,
    pub name: String,
}

impl Default for FirmwareImage {
    fn default() -> Self {
        FirmwareImage {
            path: None,
            data: Vec::new(),
            binary_type: BinaryType::Unknown,
            checksum: "0".to_string(),
            entropy: 0.0,
            config_regions: vec![],
        }
    }
}

/// Result of comparing two firmware images
#[derive(Debug)]
pub struct DiffResult {
    pub new_binaries: Vec<BinaryDiff>,
    pub modified_binaries: Vec<ModifiedRegion>,
    pub flipped_flags: Vec<ConfigFlagChange>,
    pub added_certs: Vec<CertInfo>,
    pub shifted_entropy: Vec<EntropyShift>,
}

/// Information about a newly detected binary
#[derive(Debug)]
pub struct BinaryDiff {
    pub path: String,
    pub offset: u64,
    pub size: usize,
    pub new_type: BinaryType,
    pub old_type: Option<BinaryType>,
}

/// A region that was modified between images
#[derive(Debug)]
pub struct ModifiedRegion {
    pub offset: u64,
    pub size: usize,
    pub description: String,
}

/// A configuration flag that changed state
#[derive(Debug)]
pub struct ConfigFlagChange {
    pub name: String,
    pub old_value: bool,
    pub new_value: bool,
    pub offset: u64,
}

/// Information about a detected certificate
#[derive(Debug)]
pub struct CertInfo {
    pub offset: u64,
    pub size: usize,
    pub subject: String,
    pub issuer: String,
}

/// A shift in high-entropy regions between images
#[derive(Debug)]
pub struct EntropyShift {
    pub old_offset: Option<u64>,
    pub new_offset: u64,
    pub size: usize,
    pub description: String,
}

/// Magic bytes for common binary types
const MAGIC_BYTES: &[(&[u8], BinaryType)] = &[
    (b"\x7fELF", BinaryType::Elf32),
    (b"\x7f454c46", BinaryType::Elf64), // Little endian ELF 64
    (b"\x4e4c467f", BinaryType::Elf64), // Big endian ELF 64
    (b"MZ", BinaryType::Pe32),
    (b"PE\x00\x00", BinaryType::Pe64),
];

/// Detect the binary type from raw data
pub fn detect_binary_type(data: &[u8]) -> BinaryType {
    if data.len() < 4 {
        return BinaryType::Unknown;
    }

    for (magic, btype) in MAGIC_BYTES.iter().copied() {
        if data[..magic.len()] == magic {
            return *btype;
        }
    }

    // Check for FAT filesystem signatures
    if data.len() >= 512 && data[0x3e] == b'F' && data[0x3f] == b'A' {
        return BinaryType::Fat;
    }

    BinaryType::Unknown
}

/// Calculate Shannon entropy of a byte slice
pub fn calculate_entropy(data: &[u8]) -> f64 {
    if data.is_empty() {
        return 0.0;
    }

    let mut freq = [0u32; 256];
    for &b in data.iter().take(1024) { // Limit to first 1KB for speed
        freq[b as usize] += 1;
    }

    let total: u32 = freq.iter().sum();
    if total == 0 {
        return 0.0;
    }

    let mut entropy = 0.0;
    for &count in &freq {
        if count > 0 {
            let prob = count as f64 / total as f64;
            entropy -= (prob * prob.ln()).max(0.0);
        }
    }

    // Normalize to max of 8.0 for bytes
    entropy / 8.0
}

/// Read a firmware image from disk
pub fn read_firmware(path: &Path) -> io::Result<FirmwareImage> {
    let mut data = Vec::new();
    fs::read(path)?;
    
    if data.is_empty() {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "Empty firmware image",
        ));
    }

    let binary_type = detect_binary_type(&data);
    let checksum = format!("{:x}", Sha256::digest(&data));
    let entropy = calculate_entropy(&data);

    // Default config regions for common types
    let mut config_regions = Vec::new();
    
    match binary_type {
        BinaryType::Elf32 | BinaryType::Elf64 => {
            // ELF has various headers and sections that might contain config
            if data.len() > 0x100 {
                // Check for .config section or similar
                let mut found_config = false;
                for &section in &[b".config", b"CONFIG", b"CFG"] {
                    if let Ok(offset) = find_section(&data, section) {
                        config_regions.push(ConfigRegion {
                            offset: (offset + 64) as u64, // After header
                            size: (data.len() - 0x100).min(0x1000),
                            name: format!("{} section", std::str::from_utf8(section).unwrap_or("config")),
                        });
                        found_config = true;
                    }
                }
            }
        }
        BinaryType::Pe32 | BinaryType::Pe64 => {
            // PE headers often have resource sections with config
            if data.len() > 0x180 {
                let image_size: u32 = u32::from_le_bytes([data[0x18], data[0x19], data[0x1a], data[0x1b] as u32]);
                if image_size > 0 && (image_size as usize) < data.len() {
                    // Check for common resource config regions
                    let mut found = false;
                    for &res in &[b"RCDATA", b".rsrc", b"CONFIG"] {
                        if let Ok(offset) = find_section(&data, res) {
                            config_regions.push(ConfigRegion {
                                offset: (offset + 64) as u64,
                                size: (data.len() - 0x180).min(0x2000),
                                name: format!("{} resource", std::str::from_utf8(res).unwrap_or("resource")),
                            });
                            found = true;
                        }
                    }
                }
            }
        }
        _ => {}
    }

    Ok(FirmwareImage {
        path: Some(path.to_path_buf()),
        data,
        binary_type,
        checksum,
        entropy,
        config_regions,
    })
}

/// Find a section by name in the image (simplified search)
fn find_section(data: &[u8], name: &[u8]) -> io::Result<usize> {
    let mut offset = 0;
    while offset + name.len() <= data.len() {
        if &data[offset..offset + name.len()] == name {
            return Ok(offset);
        }
        offset += 4; // Skip by 4 bytes
    }
    Err(io::Error::new(
        io::ErrorKind::NotFound,
        format!("Section not found: {:?}", std::str::from_utf8(name).unwrap_or("???")),
    ))
}

/// Compare two firmware images and produce a detailed diff report
pub fn compare_binaries(img1: &FirmwareImage, img2: &FirmwareImage) -> DiffResult {
    let mut result = DiffResult::default();

    // 1. Check for new binaries (files present in one but not other)
    // For simple comparison, we check if the overall checksum changed significantly
    if img1.checksum != img2.checksum {
        // Calculate similarity ratio
        let common_bytes: u32 = img1.data.iter()
            .zip(img2.data.iter())
            .filter(|(a, b)| a == b)
            .count();
        
        let total = (img1.data.len() + img2.data.len()).max(1);
        let similarity = common_bytes as f64 / total as f64;

        if similarity < 0.95 {
            // Significant change - likely new binaries or major modifications
            result.new_binaries.push(BinaryDiff {
                path: format!("root (similarity {:.2}%)", similarity * 100),
                offset: 0,
                size: img1.data.len().min(img2.data.len()),
                new_type: img2.binary_type,
                old_type: Some(img1.binary_type),
            });
        }
    }

    // 2. Check config regions for flipped flags
    let all_regions = [
        &img1.config_regions,
        &img2.config_regions,
    ]
    .iter()
    .flatten()
    .collect::<Vec<_>>();

    for region in all_regions {
        if region.name.contains("config") || region.name.contains("CFG") {
            // Compare the config region content
            let offset = (region.offset as usize).min(img1.data.len());
            let size = region.size.min(img1.data.len().saturating_sub(offset));
            
            if img2.data.len() > offset + size {
                let old_data: Vec<u8> = img1.data[offset..offset+size].to_vec();
                let new_data: Vec<u8> = img2.data[offset..offset+size].to_vec();

                // Look for single-bit flips (likely boolean flags)
                if let Some(flag_change) = detect_flag_flip(&old_data, &new_data) {
                    result.flipped_flags.push(ConfigFlagChange {
                        name: format!("{} flag", region.name),
                        old_value: flag_change.old_value,
                        new_value: flag_change.new_value,
                        offset: (region.offset + flag_change.bit_offset) as u64,
                    });
                }
            }
        }
    }

    // 3. Check for added certificates using entropy and magic bytes
    result.added_certs = detect_certificates(&img1.data, &img2.data);

    // 4. Detect shifted entropy regions
    result.shifted_entropy = detect_shifted_entropy(&img1.data, &img2.data);

    // 5. Find modified binary regions (non-trivial changes)
    result.modified_binaries = find_modified_regions(img1, img2);

    result
}

/// Detect flipped boolean flags between two data slices
fn detect_flag_flip(old_data: &[u8], new_data: &[u8]) -> Option<FlagFlip> {
    if old_data.len() != new_data.len() || old_data.is_empty() {
        return None;
    }

    // Look for single-bit differences (likely boolean flags)
    let mut bit_offset = 0;
    
    for (i, (&old_b, &new_b)) in old_data.iter().zip(new_data.iter()).enumerate() {
        if old_b != new_b {
            // This byte changed - check if it's a single-bit flip
            let xor = old_b ^ new_b;
            
            // Single bit flip means only one bit differs
            if xor.count_ones() == 1 {
                let bit_offset_in_byte = (xor.trailing_zeros()) as usize;
                
                // Return the first significant flag change found
                return Some(FlagFlip {
                    byte_offset: i,
                    bit_offset: bit_offset_in_byte,
                    old_value: old_b & (1 << bit_offset_in_byte) != 0,
                    new_value: new_data[i] & (1 << bit_offset_in_byte) != 0,
                });
            }
        }
    }

    None
}

/// Information about a detected flag flip
struct FlagFlip {
    byte_offset: usize,
    bit_offset: u8,
    old_value: bool,
    new_value: bool,
}

/// Detect X.509 certificates in firmware data
fn detect_certificates(old_data: &[u8], new_data: &[u8]) -> Vec<CertInfo> {
    let mut certs = Vec::new();

    // Common certificate magic bytes and sizes
    const CERT_SIZES: &[usize] = &[128, 256, 512, 1024];
    
    for (i, &size) in CERT_SIZES.iter().copied().enumerate() {
        let offset = i * size;
        
        if old_data.len() > offset + size && new_data.len() > offset + size {
            // Check if this region looks like a certificate
            // Certificates typically have high entropy and specific headers
            
            let old_entropy = calculate_entropy(&old_data[offset..offset+size]);
            let new_entropy = calculate_entropy(&new_data[offset..offset+size]);

            // Certificate-like: medium-high entropy, reasonable size
            if (0.5..1.0).contains(&old_entropy) || 
               (0.5..1.0).contains(&new_entropy) {
                
                let is_new = new_entropy > old_entropy && 
                           new_data[offset] == 0x30 && // ASN.1 SEQUENCE tag
                           new_data[offset+1] >= 2;    // Reasonable length

                if is_new || (old_entropy < 0.4) {
                    certs.push(CertInfo {
                        offset: offset as u64,
                        size,
                        subject: format!("Region at 0x{:x}", offset),
                        issuer: "Unknown".to_string(),
                    });
                }
            }
        }
    }

    // Also scan for ASN.1 SEQUENCE headers (common in PEM/DER certs)
    let mut asn1_offset = 0;
    while asn1_offset + 2 <= new_data.len() {
        if new_data[asn1_offset] == 0x30 && new_data[asn1_offset+1] >= 4 {
            // Potential SEQUENCE - check entropy of following bytes
            let chunk = &new_data[asn1_offset..(asn1_offset + 256).min(new_data.len())];
            if calculate_entropy(chunk) > 0.6 && chunk.len() < 1024 {
                certs.push(CertInfo {
                    offset: asn1_offset as u64,
                    size: chunk.len(),
                    subject: format!("ASN.1 at 0x{:x}", asn1_offset),
                    issuer: "Potential Certificate".to_string(),
                });
            }
        }
        asn1_offset += 2;
    }

    certs.dedup();
    certs
}

/// Detect regions where high-entropy content shifted between images
fn detect_shifted_entropy(old_data: &[u8], new_data: &[u8]) -> Vec<EntropyShift> {
    let mut shifts = Vec::new();

    // Find all high-entropy regions in both images
    const HIGH_ENTROPY_THRESHOLD: f64 = 0.7;
    
    fn find_high_entropy_regions(data: &[u8], window: